modules/signals/src/state-source.ts

```diff
@@ -30,10 +39,76 @@ export function patchState<State extends object>(
       currentState
     )
   );
+
+  notifyWatchers(stateSource);
 }
 
 export function getState<State extends object>(
   stateSource: StateSource<State>
 ): State {
   return stateSource[STATE_SOURCE]();
 }
+
+export function watchState<State extends object>(
+  stateSource: StateSource<State>,
+  watcher: StateWatcher<State>,
+  config?: { injector?: Injector }
+): { destroy(): void } {
+  if (!config?.injector) {
+    assertInInjectionContext(watchState);
+  }
+
+  const injector = config?.injector ?? inject(Injector);
+  const destroyRef = injector.get(DestroyRef);
+
+  addWatcher(stateSource, watcher);
+  watcher(getState(stateSource));
+
+  const destroy = () => removeWatcher(stateSource, watcher);
+  destroyRef.onDestroy(destroy);
+
+  return { destroy };
+}
+
+const STATE_WATCHERS = new WeakMap<object, Array<StateWatcher<any>>>();
+
+type StateWatcher<State extends object> = (state: NoInfer<State>) => void;
```

Comment:
Should we export this to help devs create reusable watchers? 
On the other hand it isn't a complex type...